### PR TITLE
Feature/complex methods to repository

### PR DESCRIPTION
**Pull request by luismiguel0529**
## Summary
- Add complex method in repository
- Add complex method in controller
- Add Test
## Trello Card
https://trello.com/c/Cee9M96I/25-agregar-m%C3%A9todos-m%C3%A1s-complejos-a-los-repositorios

### CODE REVIEW

File: src/main/java/wolox/training/controllers/BookController.java [hunk 1]
```diff
@@ -146,4 +158,32 @@ public void delete(@PathVariable Long id) {
         }
     }
 
+    /**
+     *Method to search a book by the following variables
```
Comment (joredudiaz) [inline src/main/java/wolox/training/controllers/BookController.java]: Can you add a space between * and the comment please?

File: src/main/java/wolox/training/controllers/BookController.java [hunk 2]
```diff
@@ -146,4 +158,32 @@ public void delete(@PathVariable Long id) {
...
+     *
+     * @param publisher variable used to create the filter
+     * @param genre variable used to create the filter
+     * @param year variable used to create the filter
+     * @return
```
Comment (joredudiaz) [inline src/main/java/wolox/training/controllers/BookController.java]: return object is missing

File: src/main/java/wolox/training/controllers/UsersController.java [hunk 3]
```diff
@@ -208,6 +210,32 @@ public void removeBook(@PathVariable Long id, @PathVariable Long bookid) {
     }
 
     /**
+     *Method to search user with birthdate beetween two date
+     *
+     * @param startDate initial date
+     * @param endDate   end date
+     * @param name name of user en repository
+     * @return
```
Comment (joredudiaz) [inline src/main/java/wolox/training/controllers/UsersController.java]: return object is missing

File: src/main/java/wolox/training/controllers/UsersController.java [hunk 4]
```diff
@@ -208,6 +210,32 @@ public void removeBook(@PathVariable Long id, @PathVariable Long bookid) {
     }
 
     /**
+     *Method to search user with birthdate beetween two date
```
Comment (joredudiaz) [inline src/main/java/wolox/training/controllers/UsersController.java]: Can you add a space between * and the comment please?

File: src/main/java/wolox/training/repositories/BookRepository.java [hunk 5]
```diff
@@ -17,4 +18,7 @@
     public Optional<Book> findByAuthor(String author);
 
     public Optional<Book> findByIsbn(String isbn);
+
+    public Optional<List<Book>> findByPublisherAndGenreAndYear(String publisher, String genre, String year);
```
Comment (joredudiaz) [inline src/main/java/wolox/training/repositories/BookRepository.java]: Can you add it java doc please?

File: src/main/java/wolox/training/repositories/UsersRepository.java [hunk 6]
```diff
@@ -15,4 +17,9 @@
 public interface UsersRepository extends JpaRepository<User, Long> {
 
     public Optional<User> findByUsername(String username);
+
+    List<User> findAllByBirthdateBetweenAndNameContainingIgnoreCase(
```
Comment (joredudiaz) [inline src/main/java/wolox/training/repositories/UsersRepository.java:35]: Can you add it java doc please?

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 7]
```diff
@@ -179,6 +175,17 @@ void whenFindBookByIsbnThenRetunrStatusCreated() throws Exception {
                 .contentType(MediaType.APPLICATION_JSON))
                 .andDo(print())
                 .andExpect(status().isCreated());
+    }
 
+    @WithMockUser(value = "miguel")
+    @Test
+    @DisplayName("Test , When a book is seached by publisher , genre and year ,it return status OK")
+    void whenFindByPublisherGenreAndYearThenReturnStatusOK() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/BookControllerTest.java:185]: You should test it when book is not found

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 8]
```diff
@@ -216,6 +223,30 @@ void whenRemoveBookThenReturnStatusNoContent() throws Exception {
...
+        String url = (USER_PATH + "/auth-username");
+        mvc.perform(get(url)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk());
+    }
+
+    @WithMockUser(value = "miguel")
+    @Test
+    @DisplayName("test, When find user by birthdate between two date, it return status OK")
+    void whenFindUserBetweenBirthdateThenReturnStatusOK() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/UserControllerTest.java:238]: You should test it when not found users

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 9]
```diff
@@ -216,6 +223,30 @@ void whenRemoveBookThenReturnStatusNoContent() throws Exception {
                 .andExpect(status().isNoContent());
     }
 
+    @WithMockUser(value = "miguel")
+    @Test
+    @DisplayName("test, When find authenticated user , it return status OK")
+    void whenFindAuthenticatedUserThenRetunrStatusOK() throws Exception {
```
Comment (joredudiaz) [inline src/test/java/wolox/training/controllers/UserControllerTest.java]: You should test it when not found user

File: src/main/java/wolox/training/controllers/BookController.java (hunk 1 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/controllers/BookController.java]: corrected

File: src/main/java/wolox/training/controllers/BookController.java (hunk 2 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/controllers/BookController.java]: corrected

File: src/main/java/wolox/training/controllers/UsersController.java (hunk 3 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/controllers/UsersController.java]: corrected

File: src/main/java/wolox/training/controllers/UsersController.java (hunk 4 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/controllers/UsersController.java]: corrected

File: src/main/java/wolox/training/repositories/BookRepository.java (hunk 5 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/repositories/BookRepository.java]: corrected

File: src/main/java/wolox/training/repositories/UsersRepository.java (hunk 6 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/repositories/UsersRepository.java:35]: corrected

File: src/test/java/wolox/training/controllers/BookControllerTest.java (hunk 7 above)
Comment (luismiguel0529) [reply src/test/java/wolox/training/controllers/BookControllerTest.java:185]: corrected

File: src/main/java/wolox/training/controllers/BookController.java [hunk 10]
```diff
@@ -145,4 +157,32 @@ public void delete(@PathVariable Long id) {
...
+            @ApiResponse(code = 200, message = "Book found successfully"),
+            @ApiResponse(code = 404, message = "Book not found")
+    })
+    @GetMapping("/{publisher}/{genre}/{year}")
+    public ResponseEntity<List<Book>> findByPublisherGenreYear(
+            @PathVariable String publisher,
+            @PathVariable String genre,
+            @PathVariable String year){
+
+        List<Book> book = bookRepository
```
Comment (facucompe) [inline src/main/java/wolox/training/controllers/BookController.java]:
```suggestion
        List<Book> books = bookRepository
```

File: src/main/java/wolox/training/controllers/BookController.java [hunk 11]
```diff
@@ -145,4 +157,32 @@ public void delete(@PathVariable Long id) {
...
+                .findByPublisherAndGenreAndYear(publisher,genre,year)
+                .orElseThrow(BookNotFoundException::new);
+
+        return new ResponseEntity<>(book,HttpStatus.OK);
```
Comment (facucompe) [inline src/main/java/wolox/training/controllers/BookController.java]:
```suggestion
        return new ResponseEntity<>(books,HttpStatus.OK);
```

**Review by facucompe — DISMISSED** [review]

File: src/test/java/wolox/training/controllers/UserControllerTest.java (hunk 8 above)
Comment (luismiguel0529) [reply src/test/java/wolox/training/controllers/UserControllerTest.java:238]: corrected

File: src/test/java/wolox/training/controllers/UserControllerTest.java (hunk 9 above)
Comment (luismiguel0529) [reply src/test/java/wolox/training/controllers/UserControllerTest.java]: when the user is not authenticated, the security component responds with a 401 Unauthorized

**Review by joredudiaz — DISMISSED** [review]

**Review by joredudiaz — DISMISSED** [review]

File: src/main/java/wolox/training/repositories/BookRepository.java [hunk 12]
```diff
@@ -14,7 +15,30 @@
...
+
+    /**
+     * Method to search book by isbn or genre or year
+     *
+     * @param publisher variable to search object
+     * @param genre     variable to search object
+     * @param year      variable to search object
+     * @return return a book with specified parameters
+     */
+    public Optional<List<Book>> findByPublisherAndGenreAndYear(String publisher, String genre, String year);
```
Comment (facucompe) [inline src/main/java/wolox/training/repositories/BookRepository.java]: In this case do not use optional of list, just list

File: src/main/java/wolox/training/repositories/UsersRepository.java [hunk 13]
```diff
@@ -14,5 +16,24 @@
...
+
+    /**
+     * Method to search users by startDate or endDate or name
+     *
+     * @param startDate variable to search object
+     * @param endDate   variable to search object
+     * @param name      variable to search object
+     * @return return a user with specified parameters
+     */
+    Optional<List<User>> findAllByBirthdateBetweenAndNameContainingIgnoreCase(
```
Comment (facucompe) [inline src/main/java/wolox/training/repositories/UsersRepository.java]: Same here

**Review by facucompe — CHANGES_REQUESTED** [review]
just that correction

**Review by joredudiaz — APPROVED** [review]

**Review by facucompe — APPROVED** [review]

**Pull request merged**